src/cargo/ops/resolve.rs

```diff
@@ -159,7 +173,7 @@ fn resolve_with_registry<'cfg>(
         true,
     )?;
 
-    if !ws.is_ephemeral() {
+    if !ws.is_ephemeral() && ws.require_optional_deps() {
```

Comment:
I was initially curious as to what's going on here, but this seems like it's connected to weirdness around `cargo install`.

Honestly I've entirely lost track of why `cargo install` is so weird at this point. IIRC though `require_optional_deps` was a `-Z` flag as well we want to have work on some day, so this raises a bit of a red flag with me in that it could be something we accidentally forget is connected to the `-Z` flag perhaps?

Reply:
Yea it is funky, and I have to look it up every time I come across it.

`cargo install` disables `require_optional_deps` so that people who mirror a registry (like with a directory registry) don't need to have *every* optional dependency on disk (only the enabled ones). You can see the original motivation in #3369.

`-Zavoid-dev-deps` is another way to disable `require_optional_deps`, to get equivalent behavior with other commands, and was added much later.

I'm pretty sure we do not want to write `Cargo.lock` in this mode, since it will be incomplete, and I think this has just been a pre-existing bug. But nothing really used it, so it didn't really matter or get noticed. 

I changed it so that I could use `resolve_ws` in the "yank" check for `cargo install`, because I didn't want to trigger downloads, and didn't need the full power of `resolve_ws_with_opts`. There are some comments above that I removed explaining this.

Reply:
Ok that sounds reasonable enough. Not great I think for the future of `-Zavoid-dev-deps`, but hey that's a problem for another day.